server/stream.go

```diff
@@ -6069,18 +6069,33 @@ func (mset *stream) ackMsg(o *consumer, seq uint64) {
 	case InterestPolicy:
 		shouldRemove = mset.noInterest(seq, o)
 	}
-	mset.mu.Unlock()
 
 	// If nothing else to do.
 	if !shouldRemove {
+		mset.mu.Unlock()
 		return
 	}
 
-	// If we are here we should attempt to remove.
-	if _, err := store.RemoveMsg(seq); err == ErrStoreEOF {
-		// This should not happen, but being pedantic.
-		mset.registerPreAckLock(o, seq)
+	if !mset.isClustered() {
+		mset.mu.Unlock()
+		// If we are here we should attempt to remove.
+		if _, err := store.RemoveMsg(seq); err == ErrStoreEOF {
+			// This should not happen, but being pedantic.
+			mset.registerPreAckLock(o, seq)
+		}
+		return
 	}
+
+	// Only propose message deletion to the stream if we're consumer leader,
```

Comment:
Stream leader?

Reply:
Could be either, the important part here is that we don't duplicate proposing message deletes.
But doing so from the stream leader would reduce required traffic, so changed it.

Reply:
Turns out this must be the consumer leader. As only the consumer leader doesn't go into the 'register pre-ack' path. And if we use the stream leader we might not have applied the message that got acked, register as pre-ack, and then not be able to propose message deletion.